doc/source/data/performance-tips.rst

```diff
@@ -62,11 +167,160 @@ For example, use ``ray.data.read_parquet("s3://anonymous@ray-example-data/iris.p
+    print(ds.materialize().stats())
+    # Stage 1 ReadRange->MapBatches(<lambda>): 7/7 blocks executed in 1.08s
+    # ...
+    # * Peak heap memory usage (MiB): 587.09 min, 1569.57 max, 1207 mean
+    # * Output num rows: 40 min, 160 max, 142 mean, 1000 total
+    # * Output size bytes: 40000160 min, 160000640 max, 142857714 mean, 1000004000 total
+    # ...
+
+Improving heap memory usage in Ray Data is an active area of development.
+Here are the current known cases in which heap memory usage may be very high:
+
+1. Reading large (1GiB or more) binary files.
+2. Transforming a Dataset where individual rows are large (100MiB or more).
+
+In these cases, the last resort is to reduce the number of concurrent execution slots.
+This can be done via custom resources.
+For example, use :meth:`ds.map_batches(fn, num_cpus=2) <ray.data.Dataset.map_batches>` to halve the number of execution slots for the ``map_batches`` tasks.
+
+If these strategies are still insufficient, please `file a Ray Data issue on GitHub`_!
+
+
+Avoiding object spilling
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+A Dataset's intermediate and output blocks are stored in Ray's object store.
+Although Ray Data attempts to minimize object store usage via :ref:`streaming execution <streaming_execution>`, it is still possible that the working set exceeds the object store capacity.
+In this case, Ray begins spilling blocks to disk, which can slow down execution significantly.
+
+There are some cases where spilling is expected. In particular, if the total Dataset's size is larger than object store capacity, and one of the following is true:
+
+1. An :ref:`all-to-all shuffle operation <optimizing_shuffles>` is used. Or,
+2. There is a call to :meth:`ds.materialize() <ray.data.Dataset.materialize>`.
+
+Otherwise, it is best to tune your application to avoid spilling.
+The recommended strategy is to manually increase the :ref:`read parallelism <read_parallelism>` or modify your application code to ensure that each task reads a smaller amount of data.
+
+.. note:: This is an active area of development. If your Dataset is causing spilling and you don't know why, `file a Ray Data issue on GitHub`_.
+
+Handling too-small blocks
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When different stages of your Dataset produce different-sized outputs, you may end up with very small blocks, which can hurt performance due to higher system overheads.
+Use :meth:`ds.stats() <ray.data.Dataset.stats>` to ensure that each stage's output blocks are each at least 1MB and ideally 100MB.
+
+If your blocks are smaller than this, consider repartitioning into larger blocks.
+There are two ways to do this:
+
+1. If you need control over the exact number of output blocks, use :meth:`ds.repartition(num_partitions) <ray.data.Dataset.repartition>`. Note that this is an :ref:`all-to-all operation <optimizing_shuffles>` and will materialize all blocks into memory before performing the repartition.
```

Comment:
Intuitively, I feel like materializing all of the data in memory would hurt performance more than having too-small blocks. Is that not the case?

Reply:
Depends on the use case. But in general, users should not expect Datasets to work well / complete if blocks are <1MiB.